feat: add internalization, add chinese translations, add missions column, add partial missions

ckx000 opened the pull request:
Added multilingual support (currently Simplified Chinese added)
Fixed a setting error
Modified and improved the handling of mission kill target completion
Added more information display (better tracking of completion progress)
Optimized layout display

AlphaConqueror committed:
Implemented feature using 'MissionRedirected': missions with completed objectives will now be removed. Program will also filter out finished missions without relying on real time events.

AlphaConqueror committed:
Removed commander dependency completely since mission ids are unique.

ckx000 committed:
updata info

ckx000 committed:
fix & updata L10n

ckx000 committed:
Create zh-Hans.strings

ckx000 committed:
add iscompleted

ckx000 committed:
is_completed tag

ckx000 committed:
add iscomplete fun

ckx000 committed:
add L10n & add displayed information

ckx000 committed:
Update README.md

ckx000 committed:
更新双语显示

ckx000 committed:
updata english commit

ckx000 committed:
undo

ckx000 committed:
updata readme

ckx000 committed:
updata pic

CMDR-WDX commented:
I ill try to review this tonight :)

CMDR-WDX commented on version:
```diff
@@ -1 +1 @@
-2.0.5
+2.1.0
```
can you please undo the version change. I want to add German translations after this is merged :)

ckx000 replied:
Of course, we can delay merging this change until you add German support. As for GitHub operations, I'm not entirely clear, but it seems possible to modify directly within this pull request?

CMDR-WDX replied:
hmm… yeah. That's a good point. I think I can directly push onto this branch. Alternatively, I can create a `develop` branch and we merge it into there :D

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -325,6 +375,7 @@ def set_frame(self, frame: tk.Frame):
         if cspan < 1:
             cspan = 2
         self.__frame = tk.Frame(frame)
+        #self.__frame.config(bg="red") # debug
```
```suggestion
```

ckx000 replied:
This commented-out code was used to display the debug UI layout. Are you suggesting that such debug code should not be retained in the codebase either?

CMDR-WDX replied:
yes. Commented out code is code smell. It's fine during development, but should not end up in "production code"

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -204,35 +231,54 @@ def __display_row(frame: tk.Frame, faction: str, data: MassacreMissionData.Facti
+    reward_sum = tk.Label(frame, text=f"{reward_sum_normal} ({reward_sum_wing})")
+    for i, entry in enumerate([label, cmp_num, kill_sum, reward_sum]):
+        entry.config(fg="YellowGreen")
+        entry.grid(row=row, column=i, sticky=tk.W)
```
please also use the same column ordering like the Sum column. It might even make sense to have the sum be part of the previous element and just use a newline.

Otherwise the ordering here is a bit confusing
![image](https://github.com/user-attachments/assets/b429b0d7-fd65-4748-9d69-7fea2310f6cf)

CMDR-WDX replied:
actually… what about merging those?

Instead of 
```
CompletedSum   X   X   X.X (X.X)
AcceptedSum      X   X   X.X (X.X)
```

``` 
Finished/All ∑ X/X X/X X.X (X.X)/
                       X.X (X.X) 
```

ckx000 replied:
I previously tried debugging this display method, but there's an issue: if forced to display in a single row, it stretches the grid layout excessively wide, causing awkward spacing in the data above.
And putting "Total Rewards" on a separate line feels like something is missing.

The intentional left/right alignment of these two rows was designed to mimic an "x/x" display style without the stretching caused by forcing single-row layout. (Apologies if the English doesn’t fully convey my intent.) Alternatively, should both rows be right-aligned?

CMDR-WDX replied:
Yeah. I think right-aligned would be the best to make it consistent with the rest

PS: no need to apologize :) You're doing great!

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -204,35 +231,54 @@ def __display_row(frame: tk.Frame, faction: str, data: MassacreMissionData.Facti
     """
-    label = tk.Label(frame, text="Sum")
+    label = tk.Label(frame, text=_("AcceptedSum"))
+    allmissionnum = tk.Label(frame, text=data.mission_count)
```
```suggestion
    all_missions_num = tk.Label(frame, text=data.mission_count)
```

ckx000 replied:
ok

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -179,16 +203,19 @@ def __display_data_header(frame: tk.Frame, settings: GridUiSettings, row=0):
+    for col in range(3+ int(settings.delta)):  # Adjust the number of columns based on whether the delta column is include
+        frame.grid_columnconfigure(col,weight=0)
+    frame.grid_columnconfigure(0,minsize=120, weight=1) 
+    # todo Check if a toggle button can be inserted here after the faction name.
```
```suggestion
```
I wouldn't add a toggle. There is little gain to see the individual missions

ckx000 replied:
This is a feature I plan to add later:
It will list details of each mission under its faction row (e.g., rewards, target counts) to help identify the lowest-value ones for optimization.

Current Pain Point:
In practice, I have to manually check the in-game mission panel to see which tasks I’ve accepted under each faction—then decide whether to abandon the lowest-reward ones for higher alternatives. This is cumbersome.

Solution:
A toggle button to:
1.Display all accepted missions per faction.
2.Highlight the lowest/highest rewards.
3.Streamline mission management.

This button's displayed information is typically used to optimize stacked missions (to maximize rewards, e.g., total rewards exceeding 500M—my friends all chase peak efficiency).
Our general workflow:
1.Accept 20/20 missions first.(Set a minimum target: e.g., only accept missions worth 20M+.)
2.Evaluate optimization potential (e.g., wait for the next 10-minute refresh to check if higher 30M+ missions are available to replace low-reward tasks).
3.The minor reputation loss from abandoning a few missions is acceptable

CMDR-WDX replied:
Hmm. Thanks for mentioning the use-case. Can you write an issue for it?
 I will then try to draw some mockups. The current UI logic is a pain point for me and I wanted to improve it for some time :D

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -82,22 +90,35 @@ def __init__(self, massacre_state: dict[int, MassacreMission]):
         """
         How many (massacre) missions does the user currently have.
         """
+        self.cmpmission_count = 0 # Add the total number of completed missions.
+        self.cmpreward = 0 # Add the total rewards from completed missions.
+        self.cmpshareable_reward = 0 # Add the total wing mission rewards from completed missions.
+        self.killcmpcount = 0 # Add the total completed kills.
```
please use snake for these. Also write out `completed` because `cmp` usually means "compare". e.g. `completed_mission_count`, `completed_reward`, etc

Also thanks for translating the comments back to english :)

ckx000 replied:
It seems I’ll need to edit the files separately—I thought changes could be made directly within this pull request.

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -19,6 +22,10 @@ class MassacreMissionData:
     @dataclass
     class FactionState:
+        
+        allnum: int # Add the number of missions for this faction.
```
```suggestion
        all_mission_count: int # Add the number of missions for this faction.
```

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -19,6 +22,10 @@ class MassacreMissionData:
     class FactionState:
+        
+        allnum: int # Add the number of missions for this faction.
+        cmpnum: int # Number of completed missions for this faction
```
```suggestion
        completed_mission_count: int # Number of completed missions for this faction
```

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -19,6 +22,10 @@ class MassacreMissionData:
+        
+        allnum: int # Add the number of missions for this faction.
+        cmpnum: int # Number of completed missions for this faction
+        killcmpcount: int # Number of completed kills for this faction
         killcount: int
```
```suggestion
        completed_kill_count: int # Number of completed kills for this faction
        kill_count: int
```

CMDR-WDX commented on massacre/ui.py:
```diff
@@ -179,16 +203,19 @@ def __display_data_header(frame: tk.Frame, settings: GridUiSettings, row=0):
+    frame.grid_columnconfigure(0,minsize=120, weight=1) 
+    # todo Check if a toggle button can be inserted here after the faction name.
+    faction_label = tk.Label(frame, text=_("Faction"))
+    missionnum_label = tk.Label(frame, text=_("R/T")) # Add the number of missions.
```
I wonder which  value seeing how many missions one has for one faction, and how many of them are already done, brings.

I can see the value for Kills, but not for missions. Could we make this an optional column that can be added in the settings?

ckx000 replied:
There are some special cases:
When the scheduled time with friends is approaching but your missions aren’t yet complete, you’ll need to evaluate:
Whether to ask friends to wait while you finish the remaining kills.
Or discuss:"I’ve currently completed X missions, totaling only XX rewards—can we share these now?"

ckx000 replied:
From my months of playing with friends, I’ve noticed they often run into this situation.
Since they don’t use plugins (or use outdated ones), they’re rarely aware of their current mission progress.

CMDR-WDX replied:
yes. I understand that. What I am trying to say is that looking at `Missions: 3/5` and `Kills: 40/60` gives you the same information, if not even more because only completed missions will show up there.

CMDR-WDX reviewed (changes requested):
General changes look good. I left some minor comments here and there. 

I would ask you not to make a new version just yet. After the merge I would also add German translations and extend the Settings a bit.

I would afterwards add another option to Hide the Info about partially finished missions because I think some people could be overwhelmed with the new information.

ckx000 committed:
Update massacre/ui.py

snake_case

Update massacre/ui.py

snake_case

Update massacre/ui.py

snake_case

Update massacre/ui.py

snake_case

Update massacre/ui.py

snake_case

Co-Authored-By: CMDR-WDX <83889860+CMDR-WDX@users.noreply.github.com>

CMDR-WDX commented:
@ckx000  I am currently adding some changes in the settings and in the UI. Could you please translate the following keys for me to Chinese? :) 

`Display partially completed missions/rewards`  - this is used in the settings
`Display Mission Count Column` - this is used in the settings

I also replaced the the  column names from "R/T" to "Missions" and "KRM/REQ" to "Kills" again as the shorthands are confusing. 

Could you then please also translate `Missions` and  `Kills`?

CMDR-WDX commented:
Github doesnt let me push onto the branch, even though `Maintainers are allowed to edit this pull request. `  is set. 

Can you apply this patch for me? 

`0001-feat-add-german-i18n-refactor-UI-code-add-new-featur.patch`
```patch
From a7facfe9e450805ab8c9cb85b2366320074a1fc3 Mon Sep 17 00:00:00 2001
From: CMDR-WDX <83889860+CMDR-WDX@users.noreply.github.com>
Date: Thu, 24 Apr 2025 00:30:08 +0200
Subject: [PATCH] feat: add german i18n, refactor UI code, add new features
 behind settings flags

---
 L10n/de.strings                          |  66 ++++++
 massacre/integrations/overlay/overlay.py |   8 +-
 massacre/massacre_settings.py            |  45 +++-
 massacre/ui.py                           | 285 ++++++++++++++++-------
 4 files changed, 301 insertions(+), 103 deletions(-)
 create mode 100644 L10n/de.strings

diff --git a/L10n/de.strings b/L10n/de.strings
new file mode 100644
index 0000000..fb89190
--- /dev/null
+++ b/L10n/de.strings
@@ -0,0 +1,66 @@
+/* Language name */
+"!Language" = "Deutsch";
+
+"Multiple Target Factions" = "Mehrere Zielfraktionen";
+
+"Multiple Target Types" = "Mehrere Zieltypen";
+
+"Multiple Target Systems" = "Mehrere Zielsysteme";
+
+"Missing Active Mission Data" = "Es Fehlen Informationen zu aktiven Missionen";
+
+"If you are in game, go to main menu and come back" = "Falls du schon im Spiel bist, gehe ins Hauptmenü und komme danach wieder zurück";
+
+"Faction" = "Fraktion";
+
+"Reward (Wing)" = "Belohnung (Wing)";
+
+"Δmax" = "Δmax";
+
+"Ratio" = "Verhältnis";
+
+"Reward" = "Belohnung";
+
+"M CR/Kill." = "M CR/Abschuss";
+
+"Kills" = "Abschüsse";
+
+"Mission Count" = "Missionenanzahl";
+
+"Massacre Plugin is Outdated" = "Massacre-Plugin ist veraltet";
+
+"Go to Download" = "Zum Download gehen";
+
+"Dismiss" = "Ausblenden";
+
+"Massacre Plugin is ready." = "Massacre Plugin ist bereit.";
+
+"UI Settings" = "UI Einstellungen";
+
+"Display Delta-Column" = "Delta-Spalte anzeigen";
+
+"Display Sum-Row" = "Summenzeile anzeigen";
+
+"Display Summary-Row" = "Zusammenfassungszeile anzeigen";
+
+"Display Mission Count" = "Missionenanzahl anzeigen";
+
+"Other" = "Sonstiges";
+
+"Check for Updates on Start" = "Beim Start nach Aktualisierungen suchen";
+
+"Made by" = "Entwickelt von";
+
+"and contributors" = "und Mitwirkenden";
+
+"Open Releases on Github" = "Versionen auf Github öffnen";
+
+"Display kills/rewards for already completed completed missions" = "Zeige Abschüsse / Belohnungen für fertige Missionen an";
+
+"Display Mission Count Summary" = "Zeige Missionenanzahl-Zeile an";
+
+"Completed Missions" = "Fertige Missionen";
+
+"All Missions" = "Alle Missionen";
+
+"Missions" = "Missionen";
\ No newline at end of file
diff --git a/massacre/integrations/overlay/overlay.py b/massacre/integrations/overlay/overlay.py
index b738549..4d59dfa 100644
--- a/massacre/integrations/overlay/overlay.py
+++ b/massacre/integrations/overlay/overlay.py
@@ -19,10 +19,10 @@ def __display_row(faction: str, data: MassacreMissionData.FactionState):
     """
     Draw one Data-Row for the Table
     """
-    reward_str = "{:.1f}".format(float(data.reward) / 1_000_000)
-    shareable_reward_str = "{:.1f}".format(float(data.shareable_reward) / 1_000_000)
+    reward_str = "{:.1f}".format(float(data.reward.total) / 1_000_000)
+    shareable_reward_str = "{:.1f}".format(float(data.shareable_reward.total) / 1_000_000)
     whole_reward_str = f'{reward_str} ({shareable_reward_str})'
-    overlay_elements = [f'{data.killcount:5}', f'{whole_reward_str:{len("Reward (Wing)")}}', f'{faction:15}']
+    overlay_elements = [f'{data.kills.total:5}', f'{whole_reward_str:{len("Reward (Wing)")}}', f'{faction:15}']
 
     return overlay_elements
 
@@ -71,7 +71,7 @@ class Overlay:
         self._create_overlay()
                 
     def __bool__(self):
-        return self.__overlay != None
+        return self.__overlay is not None
 
     def rebuild_settings(self):
         self._create_overlay()
diff --git a/massacre/massacre_settings.py b/massacre/massacre_settings.py
index 81c327a..73fe071 100644
--- a/massacre/massacre_settings.py
+++ b/massacre/massacre_settings.py
@@ -76,6 +76,21 @@ class Configuration:
         config.set(f"{self.plugin_name}.display_mission_count", value)
 
     #######################################
+    @property
+    def display_missions_column(self):
+        return config.get_bool(f"{self.plugin_name}.display_missions_column", default=False)
+
+    @display_missions_column.setter
+    def display_missions_column(self, value: bool):
+        config.set(f"{self.plugin_name}.display_missions_column", value)
+    #######################################
+    @property
+    def display_partial_stacks(self):
+        return config.get_bool(f"{self.plugin_name}.display_partial_stacks", default=False)
+    @display_partial_stacks.setter
+    def display_partial_stacks(self, value: bool):
+        config.set(f"{self.plugin_name}.display_partial_stacks", value)
+    #######################################
     def __init__(self, plugin_name: str):
        self.plugin_name = plugin_name
        self.config_changed_listeners: list[Callable[[Configuration], None]] = []
@@ -97,6 +112,10 @@ class Configuration:
             self.overlay_ttl = data['overlay_ttl'].get()
         if "display_mission_count" in keys:
             self.display_mission_count = data['display_mission_count'].get()
+        if "display_missions_column" in keys:
+            self.display_missions_column = data["display_missions_column"].get()
+        if "display_partial_stacks" in keys:
+            self.display_partial_stacks = data["display_partial_stacks"].get()
 
         for listener in self.config_changed_listeners:
             listener(self)
@@ -132,27 +151,33 @@ def build_settings_ui(root: nb.Notebook) -> tk.Frame:
     frame.columnconfigure(1, weight=1)
     __setting_changes.clear()
     __setting_changes["check_updates"] = \
-        tk.IntVar(value=configuration.check_updates)
+        tk.BooleanVar(value=configuration.check_updates)
     __setting_changes["display_delta_column"] = \
-        tk.IntVar(value=configuration.display_delta_column)
+        tk.BooleanVar(value=configuration.display_delta_column)
     __setting_changes["display_sum_row"] = \
-        tk.IntVar(value=configuration.display_sum_row)
+        tk.BooleanVar(value=configuration.display_sum_row)
     __setting_changes["display_ratio_and_cr_per_kill_row"] = \
-        tk.IntVar(value=configuration.display_ratio_and_cr_per_kill_row)
+        tk.BooleanVar(value=configuration.display_ratio_and_cr_per_kill_row)
     __setting_changes["display_mission_count"] = \
-        tk.IntVar(value=configuration.display_mission_count)
-
+        tk.BooleanVar(value=configuration.display_mission_count)
+    __setting_changes["display_partial_stacks"] = \
+        tk.BooleanVar(value=configuration.display_partial_stacks)
+    __setting_changes["display_missions_column"] = \
+        tk.BooleanVar(value=configuration.display_missions_column)
 
     nb.Label(frame, text=_("UI Settings"), pady=10).grid(sticky=tk.W, padx=title_offset)
     ui_settings_checkboxes = [
         nb.Checkbutton(frame, text=_("Display Delta-Column"),
                        variable=__setting_changes["display_delta_column"]),
+        nb.Checkbutton(frame, text=_("Display Mission Count Column"), variable=__setting_changes["display_missions_column"]),
         nb.Checkbutton(frame, text=_("Display Sum-Row"),
                        variable=__setting_changes["display_sum_row"]),
         nb.Checkbutton(frame, text=_("Display Summary-Row"),
                        variable=__setting_changes["display_ratio_and_cr_per_kill_row"]),
-        nb.Checkbutton(frame, text=_("Display Mission Count"),
-                        variable=__setting_changes["display_mission_count"])
+        nb.Checkbutton(frame, text=_("Display Mission Count Summary"),
+                        variable=__setting_changes["display_mission_count"]),
+        nb.Checkbutton(frame, text=_("Display kills/rewards for already completed completed missions"),
+                       variable=__setting_changes["display_partial_stacks"])
     ]
     for entry in ui_settings_checkboxes:
         entry.grid(columnspan=2, padx=checkbox_offset, sticky=tk.W)
@@ -167,8 +192,8 @@ def build_settings_ui(root: nb.Notebook) -> tk.Frame:
     massacre.integrations.main.notify_about_settings(frame)
     
     
-    nb.Label(frame, text="Made by CMDR WDX").grid(sticky=tk.W, padx=checkbox_offset)
-    HyperlinkLabel(frame, text="Github", background=nb.Label().cget("background"), url=download_url, underline=True)\
+    nb.Label(frame, text=f"{_("Made by")} CMDR WDX {_("and contributors")}").grid(sticky=tk.W, padx=checkbox_offset)
+    HyperlinkLabel(frame, text=_("Open Releases on Github"), background=nb.Label().cget("background"), url=download_url, underline=True)\
         .grid(columnspan=2, sticky=tk.W, padx=checkbox_offset)
 
     
diff --git a/massacre/ui.py b/massacre/ui.py
index e783036..4ea9943 100644
--- a/massacre/ui.py
+++ b/massacre/ui.py
@@ -4,6 +4,7 @@ import functools
 import tkinter as tk
 from typing import Optional
 from dataclasses import dataclass
+from enum import Enum
 
 import massacre.massacre_settings
 from massacre.massacre_mission_state import massacre_mission_listeners, MassacreMission
@@ -22,13 +23,63 @@ class MassacreMissionData:
 
     @dataclass
     class FactionState:
-        
-        all_mission_count: int # Add the number of missions for this faction.
-        completed_mission_count: int # Number of completed missions for this faction
-        completed_kill_count: int # Number of completed kills for this faction
-        kill_count: int
-        reward: int 
-        shareable_reward: int
+        @dataclass
+        class Entry:
+            total: int
+            completed: int
+            @property
+            def remaining(self):
+                return self.total - self.completed
+            @staticmethod
+            def default():
+                return MassacreMissionData.FactionState.Entry(0, 0)
+
+        missions: Entry
+        kills: Entry
+        reward: Entry
+        shareable_reward: Entry
+
+        @staticmethod
+        def default():
+            return MassacreMissionData.FactionState(
+                missions = MassacreMissionData.FactionState.Entry.default(),
+                kills = MassacreMissionData.FactionState.Entry.default(),
+                reward = MassacreMissionData.FactionState.Entry.default(),
+                shareable_reward = MassacreMissionData.FactionState.Entry.default(),
+            )
+
+    class ColumnType(Enum):
+        FACTION = 1
+        MISSIONS_OPT = 2
+        KILLS = 3
+        REWARD = 4
+        DELTA_OPT = 5
+
+        def get_alignment_head(self):
+            return self.get_alignment_body() # we might change this later. for now "inherit" this from the body
+
+        def get_alignment_body(self):
+            if self == MassacreMissionData.ColumnType.FACTION:
+                return tk.W
+            if self in [
+                MassacreMissionData.ColumnType.MISSIONS_OPT,
+                MassacreMissionData.ColumnType.KILLS,
+                MassacreMissionData.ColumnType.REWARD,
+                MassacreMissionData.ColumnType.DELTA_OPT,
+            ]:
+                return tk.E
+            logger.warning("unhandled Column: "+str(self))
+            return tk.W
+
+        def get_name(self):
+            lookup: dict[MassacreMissionData.ColumnType, str] = {
+                MassacreMissionData.ColumnType.FACTION: "Faction",
+                MassacreMissionData.ColumnType.MISSIONS_OPT: "Missions",
+                MassacreMissionData.ColumnType.KILLS: "Kills",
+                MassacreMissionData.ColumnType.REWARD: "Reward (Wing)",
+                MassacreMissionData.ColumnType.DELTA_OPT: "Δmax",
+            }
+            return lookup.get(self, "#MISSING COL DEF")
 
 
     def __init__(self, massacre_state: dict[int, MassacreMission]):
@@ -101,7 +152,7 @@ class MassacreMissionData:
 
             if mission_giver not in self.faction_to_count_lookup.keys():
                 """If no Mission from that Faction is known yet, it will first be initialized"""
-                self.faction_to_count_lookup[mission_giver] = MassacreMissionData.FactionState(0, 0, 0, 0, 0, 0)
+                self.faction_to_count_lookup[mission_giver] = MassacreMissionData.FactionState.default()
 
             faction_state = self.faction_to_count_lookup[mission_giver]
             """
@@ -109,20 +160,20 @@ class MassacreMissionData:
             from previous Missions from that faction, or 0,0,0 if this is the first mission.
             """
             
-            faction_state.all_mission_count += 1 # Add a count for the number of missions from this faction
+            faction_state.missions.total += 1 # Add a count for the number of missions from this faction
             if mission.is_completed: 
-                faction_state.completed_mission_count += 1 # Accumulate the count of completed missions.
-                faction_state.completed_kill_count += mission.count # Accumulate the count of completed kills.
+                faction_state.missions.completed += 1 # Accumulate the count of completed missions.
+                faction_state.kills.completed += mission.count # Accumulate the count of completed kills.
                 self.completed_reward += mission.reward # Add rewards from completed missions to the total.
                 if mission.is_wing: # Add rewards from completed wing missions to the total.
                     self.completed_shareable_reward += mission.reward
 
-            faction_state.killcount += mission.count
-            self.target_sum += mission.count 
-            faction_state.reward += mission.reward
+            faction_state.kills.total += mission.count
+            self.target_sum += mission.count
+            faction_state.reward.total += mission.reward
             # Only wing missions are considered for shareable rewards
             if mission.is_wing:
-                faction_state.shareable_reward += mission.reward
+                faction_state.shareable_reward.total += mission.reward
 
             ### Add Faction, Target Type and Target System to the list if they are not 
             ### yet present. This will be later used to generate a warning if more than 
@@ -136,17 +187,18 @@ class MassacreMissionData:
             if mission.target_system not in target_systems:
                 target_systems.append(mission.target_system)
 
-            if faction_state.killcount > self.stack_height:
-                self.stack_height = faction_state.killcount
+            if faction_state.kills.total > self.stack_height:
+                self.stack_height = faction_state.kills.total
 
         # After all Missions have been handled, iterate through the faction_to_count_lookup to calculate the Total Rewards   
         for faction_state in self.faction_to_count_lookup.values():
-            self.reward += faction_state.reward
-            self.shareable_reward += faction_state.shareable_reward
-            self.completed_mission_count += faction_state.completed_mission_count # Total count of all completed missions (aggregated per faction).
+            self.reward += faction_state.reward.total
+            self.shareable_reward += faction_state.shareable_reward.total
+            self.completed_mission_count += faction_state.missions.completed # Total count of all completed missions (aggregated per faction).
+            self.completed_shareable_reward += faction_state.shareable_reward.completed
             # Loop through each faction to find the largest total kills count and use it as the aggregated kill count for all missions
-            if faction_state.completed_kill_count > self.completed_kill_count:
-                self.completed_kill_count = faction_state.completed_kill_count
+            if faction_state.kills.completed > self.completed_kill_count:
+                self.completed_kill_count = faction_state.kills.completed
 
         # Check for Warnings
         if len(target_factions) > 1:
@@ -158,8 +210,8 @@ class MassacreMissionData:
 
         # Calculate before_stack_height
         for faction_state in self.faction_to_count_lookup.values():
-            if faction_state.killcount > self.before_stack_height and faction_state.killcount != self.stack_height:
-                self.before_stack_height = faction_state.killcount
+            if faction_state.kills.total > self.before_stack_height and faction_state.kills.total != self.stack_height:
+                self.before_stack_height = faction_state.kills.total
         if self.before_stack_height == 0:  # No other elements. All at max value.
             self.before_stack_height = self.stack_height
 
@@ -169,10 +221,22 @@ class GridUiSettings:
     Subset of the entire Configuration that focuses on which information is displayed
     """
     def __init__(self, config: Configuration):
-        self.sum = config.display_sum_row # fix error
-        self.delta = config.display_delta_column
-        self.summary = config.display_ratio_and_cr_per_kill_row
-        self.mission_count = config.display_mission_count
+        self.has_sum = config.display_sum_row # fix error
+        self.has_delta_col = config.display_delta_column
+        self.has_summary_row = config.display_ratio_and_cr_per_kill_row
+        self.has_mission_count_row = config.display_mission_count
+        self.has_missions_col = config.display_missions_column
+        self.display_partial_stacks = config.display_partial_stacks
+
+    def get_ordered_columns(self) -> list[MassacreMissionData.ColumnType]:
+        return [f for f in [
+            MassacreMissionData.ColumnType.FACTION,
+            MassacreMissionData.ColumnType.MISSIONS_OPT if self.has_missions_col else None,
+            MassacreMissionData.ColumnType.KILLS,
+            MassacreMissionData.ColumnType.REWARD,
+            MassacreMissionData.ColumnType.DELTA_OPT if self.has_delta_col else None,
+        ] if f is not None]
+
 
 
 def __get_row_width(settings: GridUiSettings) -> int:
@@ -180,10 +244,12 @@ def __get_row_width(settings: GridUiSettings) -> int:
     Return how many columns wide the Table is.
     This depends on if the delta-Column should be displayed
     """
-    if settings.delta:
-        return 5
-    return 4
-
+    col_count = 4
+    if settings.has_delta_col:
+        col_count += 1
+    if settings.has_missions_col:
+        col_count += 1
+    return col_count
 
 def _display_no_data_info(frame: tk.Frame):
     """
@@ -203,23 +269,16 @@ def __display_data_header(frame: tk.Frame, settings: GridUiSettings, row=0):
     """
     Display the Labels of the Table
     """
-    for col in range(3+ int(settings.delta)):  # Adjust the number of columns based on whether the delta column is include
-        frame.grid_columnconfigure(col,weight=0)
-    frame.grid_columnconfigure(0,minsize=120, weight=1) 
+    cols = settings.get_ordered_columns()
+    for idx, col in enumerate(cols):
+        weight = 0
+        if col == MassacreMissionData.ColumnType.FACTION:
+            weight = 1
+        frame.grid_columnconfigure(idx, weight=weight)
+        # TODO: Needed?
+        # frame.grid_columnconfigure(0, minsize=120, weight=1)
+        tk.Label(frame, text=_(col.get_name())).grid(row=row, column=idx, sticky=col.get_alignment_head())
     # todo Check if a toggle button can be inserted here after the faction name.
-    faction_label = tk.Label(frame, text=_("Faction"))
-    mission_num_label = tk.Label(frame, text=_("R/T")) # Add the number of missions.
-    kills_label = tk.Label(frame, text=_("KRM/REQ"))
-    payout_label = tk.Label(frame, text=_("Reward (Wing)"))
-
-    ui_elements = [faction_label, mission_num_label, kills_label, payout_label]
-    if settings.delta:
-        # noinspection SpellCheckingInspection
-        delta_label = tk.Label(frame, text=_("Δmax"))
-        ui_elements.append(delta_label)
-
-    for i, item in enumerate(ui_elements):
-        item.grid(row=row, column=i, sticky=tk.W)
 
 
 def __display_row(frame: tk.Frame, faction: str, data: MassacreMissionData.FactionState, max_count: int,
@@ -227,34 +286,58 @@ def __display_row(frame: tk.Frame, faction: str, data: MassacreMissionData.Facti
     """
     Draw one Data-Row for the Table
     """
-    reward_str = "{:.1f}".format(float(data.reward) / 1_000_000)
-    shareable_reward_str = "{:.1f}".format(float(data.shareable_reward) / 1_000_000)
-
-    faction_label = tk.Label(frame, text=faction)
-    completed_mission_count_sum = int(data.all_mission_count) - int(data.completed_mission_count)
-    mission_num_label = tk.Label(frame, text=f"{completed_mission_count_sum}/{data.all_mission_count}") # Add the number of missions.
-    completed_kill_sum =int(data.killcount) - int(data.completed_kill_count)
-    kills_label = tk.Label(frame, text=f"{completed_kill_sum}/{data.killcount}") # Modify the kill count display
-    payout_label = tk.Label(frame, text=f"{reward_str} ({shareable_reward_str})")
-
-    ui_elements = [faction_label, mission_num_label, kills_label, payout_label]
-    sticky_settings = [tk.W, tk.W, tk.W+tk.E , tk.W, tk.E] # Considering delta, define one more.
-    if settings.delta: 
-        # Calculate difference
-        delta = max_count - data.killcount
-        text = delta if delta > 0 else second_largest_count - max_count
-        delta_label = tk.Label(frame, text=str(text))
-        ui_elements.append(delta_label)
-        #delta_label.grid(row=row, column=4) # Next line of code processed.
-
-    for i, element in enumerate(ui_elements):
-        element.grid(row=row, column=i, sticky=sticky_settings[i])
-
-def __display_cmpsum(frame: tk.Frame, data: MassacreMissionData, _settings: GridUiSettings, row: int):
+    for i, col_type in enumerate(settings.get_ordered_columns()):
+        align = col_type.get_alignment_body()
+
+        match col_type:
+            case MassacreMissionData.ColumnType.FACTION:
+                text = faction
+
+            case MassacreMissionData.ColumnType.MISSIONS_OPT:
+                if settings.display_partial_stacks:
+                    text = f"{data.missions.completed}/{data.missions.total}"
+                else:
+                    text = str(data.missions.total)
+
+            case MassacreMissionData.ColumnType.REWARD:
+                # helper to format millions
+                def fmt(x):
+                    return f"{float(x) / 1_000_000:.1f}"
+                total_r = fmt(data.reward.total)
+                total_shareable = fmt(data.shareable_reward.total)
+                completed_r = fmt(data.reward.completed)
+                completed_share = fmt(data.shareable_reward.completed)
+
+                parts = []
+                if settings.display_partial_stacks:
+                    parts.append(f"{completed_r} ({completed_share})")
+                parts.append(f"{total_r} ({total_shareable})")
+                text = " / ".join(parts)
+
+            case MassacreMissionData.ColumnType.KILLS:
+                parts = []
+                if settings.display_partial_stacks:
+                    parts.append(str(data.kills.completed))
+                parts.append(str(data.kills.total))
+                text = "/".join(parts)
+
+            case MassacreMissionData.ColumnType.DELTA_OPT:
+                delta = max_count - data.kills.total
+                if delta > 0:
+                    text = str(delta)
+                else:
+                    text = str(second_largest_count - max_count)
+
+            case _:
+                text = "#MISSING COL DEF"
+        tk.Label(frame, text=text).grid(row=row, column=i, sticky=align)
+
+
+def __display_completed_sum(frame: tk.Frame, data: MassacreMissionData, _settings: GridUiSettings, row: int):
     """
     Add a row to display the total number of currently completed missions.
     """
-    label = tk.Label(frame, text=_("CompletedSum"))
+    label = tk.Label(frame, text=_("Completed Missions"))
     
     completed_num = tk.Label(frame, text=data.completed_mission_count) # Number of completed missions.
     kill_sum = tk.Label(frame, text=data.completed_kill_count) # Number of kills completed in missions.
@@ -266,21 +349,47 @@ def __display_cmpsum(frame: tk.Frame, data: MassacreMissionData, _settings: Grid
         entry.config(fg="YellowGreen")
         entry.grid(row=row, column=i, sticky=sticky_settings[i])
 
-def __display_sum(frame: tk.Frame, data: MassacreMissionData, _settings: GridUiSettings, row: int):
+def __display_sum(frame: tk.Frame, data: MassacreMissionData, settings: GridUiSettings, row: int):
     """
     Display the Sum-Row containing the Reward-Sum and the amount of Kills required.
     """
-    label = tk.Label(frame, text=_("AcceptedSum"))
-    all_missions_num = tk.Label(frame, text=data.mission_count)
-    kill_sum = tk.Label(frame, text=data.stack_height)
-    reward_sum_normal = "{:.1f}".format(float(data.reward) / 1_000_000)
-    reward_sum_wing = "{:.1f}".format(float(data.shareable_reward) / 1_000_000)
-    reward_sum = tk.Label(frame, text=f"{reward_sum_normal} ({reward_sum_wing})")
-    sticky_settings = [tk.W, tk.E, tk.E, tk.E]
-    for i, entry in enumerate([label, all_missions_num, kill_sum, reward_sum]):
-        entry.config(fg="green")
-        entry.grid(row=row, column=i, sticky=sticky_settings[i])
 
+    for i, col_type in enumerate(settings.get_ordered_columns()):
+        align = col_type.get_alignment_body()
+
+        match col_type:
+            case MassacreMissionData.ColumnType.FACTION:
+                text = _("Summary")
+
+            case MassacreMissionData.ColumnType.MISSIONS_OPT:
+                if settings.display_partial_stacks:
+                    text = f"{data.completed_mission_count}/{data.mission_count}"
+                else:
+                    text = str(data.mission_count)
+
+            case MassacreMissionData.ColumnType.REWARD:
+                # helper to format millions
+                def fmt(x):
+                    return f"{float(x) / 1_000_000:.1f}"
+                parts = []
+                if settings.display_partial_stacks:
+                    parts.append(f"{fmt(data.completed_reward)} ({fmt(data.completed_shareable_reward)})")
+                parts.append(f"{fmt(data.reward)} ({fmt(data.shareable_reward)})")
+                text = " / ".join(parts)
+
+            case MassacreMissionData.ColumnType.KILLS:
+                parts = []
+                if settings.display_partial_stacks:
+                    parts.append(str(data.completed_kill_count))
+                parts.append(str(data.stack_height))
+                text = "\\".join(parts)
+
+            case MassacreMissionData.ColumnType.DELTA_OPT:
+                text = ""
+
+            case _:
+                text = "#MISSING COL DEF"
+        tk.Label(frame, text=text, fg="green").grid(row=row, column=i, sticky=align)
 
 def __display_summary(frame: tk.Frame, data: MassacreMissionData, settings: GridUiSettings, row: int):
     ratio_text = "{:.2f}".format(float(data.target_sum)/float(data.stack_height))
@@ -309,18 +418,16 @@ def _display_data(frame: tk.Frame, data: MassacreMissionData, settings: GridUiSe
         row_pointer += 1
         # todo In the future, add a toggle button here to display detailed mission information under each faction.
 
-    if settings.sum:
-        __display_cmpsum(frame, data, settings, row_pointer)
-        row_pointer += 1
+    if settings.has_sum:
         __display_sum(frame, data, settings, row_pointer)
         row_pointer += 1
 
-    if settings.summary:
+    if settings.has_summary_row:
         __display_summary(frame, data, settings, row_pointer)
         row_pointer += 1
     full_width = __get_row_width(settings)
 
-    if settings.mission_count:
+    if settings.has_mission_count_row:
         __display_mission_count(frame, data, full_width, row_pointer)
         row_pointer += 1
 
-- 
2.49.0

```

You can apply it using `git am pathToPatch`

CMDR-WDX commented:
actually… i ill just merge this into a dev branch and work from there :)
